### PR TITLE
Add LogMonkey Barrel

### PR DESCRIPTION
**Pull request by KenConnectIQ**
Adds a LogMonkey barrel project that contains logging utilities
useful for debugging.

### CODE REVIEW

**Comment by blaskovicz** [comment]
If this existed I probably could have avoided writing my own logger!

https://github.com/blaskovicz/garmin-nest-camera-control/blob/master/source/Logger.mc

**Comment by blaskovicz** [comment]
@KenConnectIQ the link in your readme to 'monkey barrels' on the connect-iq developer guide doesn't seems dead; are you intending a new guide to be added there as well?
https://developer.garmin.com/connect-iq/programmers-guide/monkey-barrels

**Comment by KenConnectIQ** [comment]
I tried to update the URLs in the links in this change to where they will be before pushing. The Monkey Barrels feature is in beta right now and the documentation on the developer website reflects the current production release documentation.

I'm glad to see there's some interest in a logging tool though. I'm hoping we see enough interest in this tool to evolve and improve it over time.

File: barrels/LogMonkey/LogTests.mc [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+    // Wrap the unit tests in a module annotated with :test so it will get compiled
+    // out for non unit test builds.
+    (:test)
+    module Test {
+
+        //! This class mocks the Toybox.System.println() interface but instead
+        //! of printing a value saves it to a string value which can be retrieved
+        //! and examed to verify log results are correct.
+        class MockSys {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/LogTests.mc:15]: This class should get an `initialize()` method to avoid compiler warnings for users that enable them.

File: barrels/LogMonkey/LogTests.mc [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+            private var mStreamContents;
+
+            //! A mock of Toybox.System.println() function
+            //! @param [Toybox::Lang::Object] message The message to print/log
+            public function println(message) {
+                if (mStreamContents == null) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/LogTests.mc]: If we initialize mStreamContents to an empty string, we can avoid this conditional.

File: barrels/LogMonkey/LogTests.mc [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            }
+
+        }
+
+        // This class is used within the unit tests module to create an Exception
+        // object with a custom message.
+        class TestException extends Lang.Exception {
+            function initialize() {
+                Exception.initialize();
+                mMessage = "This is a test Exception";
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/LogTests.mc]: It appears as though you're accessing an undocumented member here. Should we instead be overriding `getErrorMessage()` to return this literal string?

File: barrels/LogMonkey/LogTests.mc [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            Test.assert(logOutput.find("(null)") != null);
+
+            // Test Lang.Array
+            varToTest = [1, 2, 3];
+            mockSys = new MockSys();
+            log = new $.LogMonkey.Logger(level, mockSys);
+            log.logVariable(tag, varName, varToTest);
+            logOutput = mockSys.getStreamContents();
+            logger.debug("log output: " + logOutput);
+            Test.assert(logOutput.find("(Array)") != null);
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/LogTests.mc]: Should the testing be more thorough and verify the actual value output matches what we expect? If so, perhaps there should be a test function for verifying each variable type we support?

File: barrels/LogMonkey/Logger.mc [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    //! is set within the Logger.initialize() function.
+    class Logger {
+
+        private var mLogLevel, mLogStream;
+
+        //! Creates a new Logger object.
+        //! @param [Toybox::Lang::String] The log level value to log messages through this class to
+        //! @param [Toybox::Lang::Object] An object which defines a println(message) function
+        (:debug)
+        function initialize(logLevel, logStream) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: Should this be documented? Are users expected to create their own loggers?

File: barrels/LogMonkey/Logger.mc [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+using Toybox.Lang as Lang;
+
+module LogMonkey {
+
+    //! The Logger class provides public APIs to log various types of
+    //! messages. Each Logger class is associated with a log level which
+    //! is set within the Logger.initialize() function.
+    class Logger {
+
+        private var mLogLevel, mLogStream;
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: Typical coding style guidelines would require each declaration to be on its own line.

File: barrels/LogMonkey/Logger.mc [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        // symbols available in the release build so we won't get symbol
+        // not found errors while saving memory.
+        (:release)
+        function initialize(logLevel, logStream) {
+        }
+
+        // The string formats to use when printing log messages
+        private static var FORMAT_VARIABLE = "$1$=($2$) $3$";
+        private static var FORMAT_LOG_MESSAGE = "(lmf1)[$1$] {$2$} $3$: $4$";
+        private static var FORMAT_TIMESTAMP = "$1$-$2$-$3$ $4$:$5$:$6$";
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: These private static are not necessary in release. They could easily be annotated out for additional savings.

File: barrels/LogMonkey/Logger.mc (hunk 6 above)
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: These private static members are not necessary in release. They could easily be annotated out for additional savings.

Comment (TravisConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Perhaps it makes sense to have two versions of this class, one annotated with :release that just declares empty versions of the public methods and another that is annotated with :debug that has all of the necessary implementation details.

File: barrels/LogMonkey/Logger.mc [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                currentTime.min.format("%02u"),
+                currentTime.sec.format("%02u")
+            ]);
+
+            // Form the log message
+            return Lang.format(FORMAT_LOG_MESSAGE, [timestamp, mLogLevel, tag, message]);
+        }
+
+        (:release)
+        private function formLogMessage(tag, message) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: This method is not needed in release. It should probably be omitted.

File: barrels/LogMonkey/Logger.mc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        //! @param [Toybox::Lang::String] tag The tag to apply to the log message
+        //! @param [Toybox::Lang::String] message The message to log
+        (:debug)
+        private function log(tag, message) {
+            // Print the log message
+            mLogStream.println(formLogMessage(tag, message));
+        }
+
+        (:release)
+        private function log(tag, message) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: This method is not needed in release. It should probably be omitted.

File: barrels/LogMonkey/Logger.mc [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                    return "Symbol";
+                case instanceof Lang.WeakReference:
+                    return "WeakReference";
+                default:
+                    return "Object";
+            }
+        }
+
+        (:release)
+        private function getVariableType(variable) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: This method is not needed in release. It should probably be omitted.

File: barrels/LogMonkey/Logger.mc [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        }
+
+        //! Log the given message under the given tag
+        //! @param tag The tag to apply to the log message
+        //! @param message The message to log
+        (:debug)
+        public function logMessage(tag, message) {
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: All MonkeyC code I've seen to date does not explicitly label public functions as such. Should this code reflect existing convention, or is it starting a revolution?

File: barrels/LogMonkey/Logger.mc [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        //! Forms a log message based on the given values.
+        //! @param [Toybox::Lang::String] tag The tag to apply to the log message
+        //! @param [Toybox::Lang::String] message The message to log
+        //! @return [Toybox::Lang::String] A log message matching the format specified by FORMAT_TIMESTAMP
+        (:debug)
+        private function formLogMessage(tag, message) {
+            // Get a timestamp from the system
+            var currentTime = Greg.info(Time.now(), Time.FORMAT_SHORT);
+            var timestamp = Lang.format(FORMAT_TIMESTAMP, [
+                currentTime.month.format("%02u"),
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/Logger.mc]: Should the timestamp format be re-arranged such that the timestamps are ordered for easy sorting? The python script included with this patch gathers log messages from various files. If this is a use case that is supported, it would make sense that users would want the log messages from various files to be joined and easily sortable.

If we ordered this as YYYY-MM-DD, the output could be sorted with the standard _sort_ utility after generation (or even internally by the python script) without having to slice the data up first.

File: barrels/LogMonkey/README.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        }
+        ...
+    }
+
+}
+```
+
+## Viewing and Parsing Logs
+
+LogMonkey makes use of `Toybox.System.println()` calls to log information. This means that if your app also makes use of `pinrtln()` calls they will show up alongside the log entries.
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: typo: `pinrtln` should be `println`

File: barrels/LogMonkey/README.md [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+See the general instructions for [including Monkey Barrels](https://github.com/garmin/connectiq-apps/tree/master/barrels#including-monkey-barrels).
+
+## Logging Information
+
+Once you've added LogMonkey to your project you can import it into your code via `using` statements.
+```
+using LogMonkey as Log;
+```
+Within you're project's code you can log messages, variables and Exceptions via the available APIs.
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: typo: `you're` should be `your`

File: barrels/LogMonkey/README.md [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+}
+```
+
+## Viewing and Parsing Logs
+
+LogMonkey makes use of `Toybox.System.println()` calls to log information. This means that if your app also makes use of `pinrtln()` calls they will show up alongside the log entries.
+
+### Output Format
+
+LogMonkey logs each line in the file in the following format: `(log format version)[timestamp] {log level} tag: message`. Details for each block of this format are given in the table below.
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: `logs each line in the file` should probably be `generates messages`

File: barrels/LogMonkey/README.md [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+| Block              | Format                                                 | Description                                                                         |
+| ------------------ | ------------------------------------------------------ | ----------------------------------------------------------------------------------- |
+| log format version | A value beginning with "lfm" then a number             | The output format version of the line in the log.                                   |
+| timestamp          | MM-DD-YYYY HH:MM:SS                                    | A time stamp of when the log entry occurred.                                        |
+| log level          | Regular Expression: `[a-zA-Z0-9_]`                     | The priority level of the log entry. LogMonkey supports values of `D`, `W` and `E`. |
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: Are log levels required to be a single alpha-numeric character or underscore? This does break into the question about whether the Logger class initialize method should be documented or if there should be an undocumented method for specifying the `level`.

File: barrels/LogMonkey/README.md [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+| tag                | A string containing any character except a colon (`:`) | A brief tag value that can be used to quickly filter and find specific log entries. |
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: The log level is described above using a regular expression, but this one in plain English. Should we be consistent?

File: barrels/LogMonkey/README.md [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+| message            | Any string value                                       | The detailed message being recorded in the log entry.                               |
+
+### Viewing Logs in Eclipse
+
+When you run an app via Eclipse the run console will show the output of log calls. You will see the information your app is logging in the raw LogMonkey output format.
+![Image of run console](run_console.png)
+
+### Viewing Logs on a Device
+
+When running your app on a physical device the log statements get added to the app's log file. The app's log file can be viewed by plugging your device into your computer and navigating to `<device root>/GARMIN/APPS/LOGS`. The app log file will match the name of the app prg file (located in `<device root>/GARMIN/APPS`.
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: Should this mention that that the user will need to create the log file in order for the logging to appear there?

File: barrels/README.md [hunk 19]
```diff
@@ -1,8 +1,25 @@
...
+your Manifest file and then link to the Barrel within your project's Jungle file. See the
+[Programmer's Guide](https://developer.garmin.com/connect-iq/programmers-guide/monkey-barrels#how-to-include-barrels)
+for a detailed description of how to include Monkey Barrels in your project.
+
+## Barrel Index
+
+### LogMonkey
+The [LogMonkey barrel](https://github.com/garmin/connectiq-apps/tree/master/barrels/LogMonkey)
+provides some basic logging utilities to aid in developing Connect IQ apps. Inclusions of the
+LogMonkey within an app will automatically be compiled out or minimized when building a release
```
Comment (TravisConnectIQ) [inline barrels/README.md]: Compiled out or minimized?

**Review by TravisConnectIQ — CHANGES_REQUESTED** [review]

File: barrels/LogMonkey/LogTests.mc (hunk 1 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/LogTests.mc:15]: Unless we use the initialize function to set mStreamContents to an empty string as suggested below - which is ultimately what I did - I don't think this is necessary. This class isn't overriding another class so there's not a parent initialize to call and the compiler will insert an empty initialize function if you don't include one so any class that would subclass this one would be able to call the initialize symbol safely.

File: barrels/LogMonkey/LogTests.mc (hunk 2 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/LogTests.mc]: Done

File: barrels/LogMonkey/LogTests.mc (hunk 3 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/LogTests.mc]: Done

File: barrels/LogMonkey/LogTests.mc (hunk 4 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/LogTests.mc]: I'd be open to looking for a more exact value if you have suggestions on how to do it. I really just wanted a test to ensure that when given a variable type to getVariableType() we'd get the expected value in parentheses in the output message. The problem I was running into is that the log format includes a timestamp so we can do an exact match and there's not a great way to do a regex type of check. We could have a function that breaks the message down and looks for a specific part to match exactly but I don't know if that will provide better value that just looking for the specific type piece.

File: barrels/LogMonkey/Logger.mc (hunk 6 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Done

Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I like the idea of separating this by class. I'm going to take that approach.

File: barrels/LogMonkey/Logger.mc (hunk 5 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: An earlier conversation with Nick resulted in restructuring the log classes with the idea that different implementations of the class could be swapped out/configured. I wanted to explicitly document what's expected as inputs to a Logger class so in the future if there's a configuration option that external developers could use to setup their own loggers or if someone internally is going in and modifying this barrel it's documented.

File: barrels/LogMonkey/Logger.mc (hunk 7 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: These should be taken care of with the empty class approach discussed above.

File: barrels/LogMonkey/Logger.mc (hunk 12 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I like this reasoning. I'll update the format.

File: barrels/LogMonkey/Logger.mc (hunk 8 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Done

File: barrels/LogMonkey/Logger.mc (hunk 9 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Done

File: barrels/LogMonkey/Logger.mc (hunk 10 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Done

File: barrels/LogMonkey/Logger.mc (hunk 11 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I think with the introduction of public/private/protected we should start using them as a coding convention even if it's not grammatically required but I'm also not the authority to dictate this. Unless there's a strong opinion against it I'm going to leave them in.

File: barrels/LogMonkey/README.md (hunk 14 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: Done

File: barrels/LogMonkey/README.md (hunk 13 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: Done

File: barrels/LogMonkey/README.md (hunk 15 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: Done

File: barrels/LogMonkey/README.md (hunk 16 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: The log level value should can be more than a single character; the regular expression here should be updated to reflect that. I wanted the standard format to be meaningful but concise so it's less overwhelming to view a raw log file.

File: barrels/LogMonkey/README.md (hunk 17 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: I think I choose to use a regular expression above because it was easier - from my view - to understand it but I'm going to use plain English for both.

File: barrels/LogMonkey/README.md (hunk 18 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: Probably. I'll add that.

File: barrels/README.md (hunk 19 above)
Comment (KenConnectIQ) [reply barrels/README.md]: I'm going to change the entire sentence to be: Including LogMonkey within an app and building the app in release mode will automatically minimize the compiled output to save run time memory.

File: barrels/LogMonkey/LogTests.mc (hunk 4 above)
Comment (TravisConnectIQ) [reply barrels/LogMonkey/LogTests.mc]: Well, MonkeyC doesn't have regular expression support, but I was thinking you could do something like this..

```
(:test)
module Test {

    // test helper function 
    function do_testLogVariable(level, tag, name, variable, match) {

        var mockSys = new MockSys();

        var log = new $.LogMonkey.Logger(level, mockSys);
        log.logVariable(tag, name, variable);
        log = null;

        var logOutput = mockSys.getStreamContents();

        // look for the caller-provided substring for an exact match
        Test.assert(logOutput.find(match) != null);
    }

    (:test)
    function testLogVariable_Number(logger) {
        do_testLogVariable("D", "x", "aNumber", -1, " {D} x: aNumber=(Number) -1");
        do_testLogVariable("W", "y", "bNumber",  0, " {W} y: bNumber=(Number) 0");
        do_testLogVariable("E", "z", "cNumber",  1, " {E} y: cNumber=(Number) 1");
        // test more values if we feel it is necessary (INT_MIN, INT_MIN + 1, INT_MAX - 1, INT_MAX)
    }

    (:test)
    function testLogVariable_Array(logger) {
        do_testLogVariable("E", "a", "xArray",           [], " {E} a: xArray=(Array) []");
        do_testLogVariable("W", "b", "yArray", [ 1, "abc" ], " {W} b: yArray=(Array) [1, abc]");
        // test more values if we feel it is necessary
    }

    (:test)
    function testLogVariable_Prefix(logger) {
        var mockSys = new MockSys();

        var log = new $.LogMonkey.Logger("I", mockSys);
        log.logVariable("i", "iI", logger);
        log = null;

        var logOutput = mockSys.getStreamContents();

        // this assert may have to change when the version changes
        // verify the version happens at the front
        Test.assert(logOutput.find("(lmf1)" == 0));
    }
}
```

If you wanted to properly handle the date/time you'd have to give the logger something that it could use to get the date/time and then mock it in the tests just as you've done with ``Sys.println()``. Given what you've got right now, the pattern would be to take two additional parameters (``Gregorian`` and ``Time``).

File: barrels/LogMonkey/LogTests.mc (hunk 1 above)
Comment (TravisConnectIQ) [reply barrels/LogMonkey/LogTests.mc:15]: You are right, the warning would only occur if there was a base.

File: barrels/LogMonkey/Logger.mc (hunk 5 above)
Comment (TravisConnectIQ) [reply barrels/LogMonkey/Logger.mc]: Technically, if users want to implement their own logger, it only needs to provide the three public log methods. I'm not sure why users would want to inherit from this logger. I could be wrong.

It just seems weird to mention the last parameter to users as this is really a nasty implementation detail that allows us to test the logger.

File: barrels/LogMonkey/Logger.mc (hunk 11 above)
Comment (TravisConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I don't mind. I only bring it up because this is just a Java convention that is sneaking into non-Java code. Ideally all of the code in this repo would follow a consistent style, but I'm not convinced that this is one of those style things that really matters.

**Review by TravisConnectIQ — COMMENTED** [review]
I'm not familiar with reviews in github, but I'd expect that you'd have to commit your fixes locally and then issue another pull request for me to review the changes. 

Is that correct?

File: barrels/LogMonkey/LogTests.mc (hunk 4 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/LogTests.mc]: I split up the test types unit test into the individual variable types using the helper function you suggested. I'm going to leave the mocked time values out as I don't want to have the unit test dictate the constructor interface for the Logger class.

File: barrels/LogMonkey/Logger.mc (hunk 5 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I don't think that's the case. The last parameter is simply stating you must provide a stream object that defines a println method. If better options for writing to a file/dedicated log get added in the future then that object can be swapped out as needed. I also think the idea was that in the future there could be some sort of setup/initialize function with LogMonkey that allows the user to define the underlying output/destination of the log which would then create the local variable objects.
```
module LogMonkey {
    var Debug;
    var Warn;
    var Error;

    function setup(logStream) {
        Debug = new Logger("D", logStream);
        Warn = new Logger("W", logStream);
        Error = new Logger("E", logStream);
    }
}
```
We're just not going to include any sort of setup now because the underlying system just doesn't support a more advanced infrastructure.

File: barrels/LogMonkey/Logger.mc (hunk 11 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/Logger.mc]: I agree it is a bit more Java-y but so is switching the grammar from hidden/<nothing> to private/protected/public. If we chose to go that route with the grammar I think it may make sense to update our teams convention approach. I'll update the patch to remove the public definitions and leave this open to debate internally.

**Comment by KenConnectIQ** [comment]
I just have to push a new change to this branch and the pull request will get updated. I haven't yet pushed changes but should be doing so before lunch today.

File: barrels/LogMonkey/README.md [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+LogMonkey makes use of `Toybox.System.println()` calls to log information. This means that if your app also makes use of `println()` calls they will show up alongside the log entries.
+
+### Output Format
+
+LogMonkey generates messages in the following format: `(log format version)[timestamp] {log level} tag: message`. Details for each block of this format are given in the table below.
+
+| Block              | Format                                                    | Description                                                                         |
+| ------------------ | --------------------------------------------------------- | ----------------------------------------------------------------------------------- |
+| log format version | A value beginning with "lfm" then a number                | The output format version of the line in the log.                                   |
+| timestamp          | MM-DD-YYYY HH:MM:SS                                       | A time stamp of when the log entry occurred.                                        |
```
Comment (TravisConnectIQ) [inline barrels/LogMonkey/README.md]: The format here needs to be updated to reflect the implementation changes.

File: barrels/LogMonkey/README.md (hunk 18 above)
Comment (TravisConnectIQ) [reply barrels/LogMonkey/README.md]: I'm not seeing anything additional here.

File: barrels/LogMonkey/README.md (hunk 20 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: Missed this one. Done.

File: barrels/LogMonkey/README.md (hunk 18 above)
Comment (KenConnectIQ) [reply barrels/LogMonkey/README.md]: I think I missed this when rebasing my change. The last commit I pushed adds this bit back in.

**Review by TravisConnectIQ — APPROVED** [review]
